litemapy/schematic.py

```diff
@@ -278,6 +278,80 @@ def _tonbt(self):
 
         return root
 
+    def to_structure_nbt(self, mc_version=MC_DATA_VERSION, gzipped=True, byteorder='big'):
+        structure = nbtlib.File(gzipped=gzipped, byteorder=byteorder)
+
+        structure['size'] = List[Int]([self.__width, self.__height, self.__length])
+        structure['DataVersion'] = Int(mc_version)
+
+        # process entities
+        entities = List[Compound]()
+        for entity in self.entities:
+            entity_cmp = Compound()
+            entity_cmp['nbt'] = entity.data
+            entity_cmp['pos'] = List[Double]([Double(coord) for coord in entity.position])
+            entity_cmp['blockPos'] = List[Int]([Int(coord) for coord in entity.position])
+            entities.append(entity_cmp)
+
+        structure['entities'] = entities
+
+        # create tile entity dictionary to add them correctly to the block list later
+        tile_entity_dict = {}
+        for tile_entity in self.tile_entities:
+            tile_entity_cmp = Compound()
+            for key, value in tile_entity.data.items():
+                if key not in ['x', 'y', 'z']:
+                    tile_entity_cmp[key] = value
+
+            tile_entity_dict[tile_entity.position] = tile_entity_cmp
+
+        # process palette
+        structure['palette'] = List[Compound]([block._tonbt() for block in self.__palette])
+
+        # process blocks
+        blocks = List[Compound]()
+        for x, y, z in self.allblockpos():
```

Comment:
`x`, `y` and `z`could be negative here, if the region has negative dimensions. I don't think that's supported by the vanilla structures. You need to make the coordinates positive. You can do so by subtracting region.minx(), region.miny() and region.minz() from x, y and z respectively.

Reply:
Tile entity positions get stored defined from the lowest, west-most, north-most (minimal values for x, y, z) corner. So they are the same as the indices of the numpy array self.__blocks. The new indexing works to correctly match the tile entities to the corresponding blocks.